src/util/container.h

```diff
@@ -303,5 +304,54 @@ class LRUCache
 	std::list<K> m_queue;
 };
 
+template <typename T, size_t N>
+class RingBuffer
+{
+public:
+	RingBuffer() : m_offset(0)
+	{
+		for (size_t i = 0; i < N; ++i)
+			m_data[i] = T(); // buffer is always full
+	}
+	
+	T& at(size_t n)
+	{
+		if (n >= N)
+			throw std::out_of_range("RingBuffer::at");
+		if (n + m_offset < N)
+			return m_data[n + m_offset];
+		else
+			return m_data[n + m_offset - N];
+	}
```

Comment:
I don't know. I think that at() should always point to the "oldest" element (this is the element that would be overridden with a push()). The ring buffer mimics a fixed size queue where you would expect the element address returned by at(0) to change. With vector, at(0) points to the first element in the sequence. So does this one. This implementation is not odd at all.